File: apps/ios/Ozer/Notes/NoteMarkdown.swift
import Foundation

enum NoteBlockKind: String, Equatable, Sendable {
    case paragraph
    case heading1
    case heading2
    case bullet
}

struct NoteInlineRun: Equatable, Sendable {
    var text: String
    var bold: Bool
    var italic: Bool
    var underline: Bool

    var isBlank: Bool {
        text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
    }
}

struct NoteBlock: Equatable, Sendable {
    var kind: NoteBlockKind
    var runs: [NoteInlineRun]

    var isBlank: Bool {
        runs.allSatisfy(\.isBlank)
    }

    var plainText: String {
        runs.map(\.text).joined()
    }
}

/// Shared markdown subset with the web TipTap notes editor:
/// bold (`**`), italic (`*`), underline (`<u>`), bullets, H1, H2.
enum NoteMarkdown {
    static func parse(_ markdown: String) -> [NoteBlock] {
        let normalized = markdown.replacingOccurrences(of: "\r\n", with: "\n")
            .replacingOccurrences(of: "\r", with: "\n")
        if normalized.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            return []
        }

        var blocks: [NoteBlock] = []
        for line in normalized.split(separator: "\n", omittingEmptySubsequences: false) {
            let raw = String(line)
            if raw.trimmingCharacters(in: .whitespaces).isEmpty {
                continue
            }
            // Keep `## Me` / `## Speaker 1` as H2 — do not flatten transcript labels.
            if raw.hasPrefix("## ") {
                blocks.append(
                    NoteBlock(kind: .heading2, runs: parseInlines(String(raw.dropFirst(3))))
                )
            } else if raw.hasPrefix("# ") {
                blocks.append(
                    NoteBlock(kind: .heading1, runs: parseInlines(String(raw.dropFirst(2))))
                )
            } else if let item = bulletItem(raw) {
                blocks.append(NoteBlock(kind: .bullet, runs: parseInlines(item)))
            } else {
                blocks.append(NoteBlock(kind: .paragraph, runs: parseInlines(raw)))
            }
        }
        return blocks
    }

    static func serialize(_ blocks: [NoteBlock]) -> String {
        let meaningful = blocks.filter { !$0.isBlank || $0.kind == .bullet }
        var parts: [String] = []
        for (index, block) in meaningful.enumerated() {
            let inline = serializeInlines(block.runs)
            let line: String
            switch block.kind {
            case .heading1:
                line = "# \(inline)"
            case .heading2:
                line = "## \(inline)"
            case .bullet:
                line = "- \(inline)"
            case .paragraph:
                line = inline
            }
            if index > 0 {
                let previous = meaningful[index - 1]
                if previous.kind == .bullet, block.kind == .bullet {
                    parts.append("\n")
                } else {
                    parts.append("\n\n")
                }
            }
            parts.append(line)
        }
        return parts.joined()
    }

    static func plainText(from markdown: String) -> String {
        parse(markdown)
            .map(\.plainText)
            .filter { !$0.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty }
            .joined(separator: "\n")
    }

    static func isBlank(_ markdown: String) -> Bool {
        plainText(from: markdown)
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .isEmpty
    }

    static func roundTrip(_ markdown: String) -> String {
        serialize(parse(markdown))
    }

    private static func bulletItem(_ line: String) -> String? {
        var rest = line[...]
        while rest.first == " " {
            rest.removeFirst()
        }
        for prefix in ["- ", "* ", "+ ", "• "] where rest.hasPrefix(prefix) {
            return String(rest.dropFirst(prefix.count))
        }
        return nil
    }

    private static func parseInlines(_ input: String) -> [NoteInlineRun] {
        merge(parseInlines(Array(input), start: 0, end: input.count, marks: Marks()).runs)
    }

    private struct Marks: Equatable {
        var bold = false
        var italic = false
        var underline = false
    }

    private struct InlineParse {
        var runs: [NoteInlineRun]
        var index: Int
    }

    private static func parseInlines(
        _ chars: [Character],
        start: Int,
        end: Int,
        marks: Marks
    ) -> InlineParse {
        var runs: [NoteInlineRun] = []
        var buffer = ""
        var index = start

        func flush() {
            guard !buffer.isEmpty else { return }
            runs.append(
                NoteInlineRun(
                    text: buffer,
                    bold: marks.bold,
                    italic: marks.italic,
                    underline: marks.underline
                )
            )
            buffer = ""
        }

        while index < end {
            if chars[index] == "\\", index + 1 < end {
                buffer.append(chars[index + 1])
                index += 2
                continue
            }

            if match(chars, index: index, end: end, token: "<u>"),
               let close = findToken(chars, token: "</u>", from: index + 3, end: end)
            {
                flush()
                var inner = marks
                inner.underline = true
                let nested = parseInlines(chars, start: index + 3, end: close, marks: inner)
                runs.append(contentsOf: nested.runs)
                index = close + 4
                continue
            }

            if let wrapped = wrap(
                chars,
                index: index,
                end: end,
                delimiter: "***",
                marks: marks,
                apply: { $0.bold = true; $0.italic = true }
            ) {
                flush()
                runs.append(contentsOf: wrapped.runs)
                index = wrapped.index
                continue
            }

            if let wrapped = wrap(
                chars,
                index: index,
                end: end,
                delimiter: "___",
                marks: marks,
                apply: { $0.bold = true; $0.italic = true }
            ) {
                flush()
                runs.append(contentsOf: wrapped.runs)
                index = wrapped.index
                continue
            }

            if let wrapped = wrap(
                chars,
                index: index,
                end: end,
                delimiter: "**",
                marks: marks,
                apply: { $0.bold = true }
            ) {
                flush()
                runs.append(contentsOf: wrapped.runs)
                index = wrapped.index
                continue
            }

            if let wrapped = wrap(
                chars,
                index: index,
                end: end,
                delimiter: "__",
                marks: marks,
                apply: { $0.bold = true }
            ) {
                flush()
                runs.append(contentsOf: wrapped.runs)
                index = wrapped.index
                continue
            }

            if let wrapped = wrap(
                chars,
                index: index,
                end: end,
                delimiter: "*",
                marks: marks,
                apply: { $0.italic = true }
            ) {
                flush()
                runs.append(contentsOf: wrapped.runs)
                index = wrapped.index
                continue
            }

            if canOpenUnderscore(chars, index: index, end: end),
               let wrapped = wrap(
                   chars,
                   index: index,
                   end: end,
                   delimiter: "_",
                   marks: marks,
                   apply: { $0.italic = true }
               )
            {
                flush()
                runs.append(contentsOf: wrapped.runs)
                index = wrapped.index
                continue
            }

            buffer.append(chars[index])
            index += 1
        }

        flush()
        return InlineParse(runs: runs, index: index)
    }

    private static func wrap(
        _ chars: [Character],
        index: Int,
        end: Int,
        delimiter: String,
        marks: Marks,
        apply: (inout Marks) -> Void
    ) -> InlineParse? {
        guard match(chars, index: index, end: end, token: delimiter) else {
            return nil
        }
        let innerStart = index + delimiter.count
        guard let close = findToken(chars, token: delimiter, from: innerStart, end: end),
              close > innerStart
        else {
            return nil
        }
        var inner = marks
        apply(&inner)
        let nested = parseInlines(chars, start: innerStart, end: close, marks: inner)
        return InlineParse(runs: nested.runs, index: close + delimiter.count)
    }

    private static func match(
        _ chars: [Character],
        index: Int,
        end: Int,
        token: String
    ) -> Bool {
        let tokenChars = Array(token)
        guard index + tokenChars.count <= end else { return false }
        for (offset, character) in tokenChars.enumerated() where chars[index + offset] != character {
            return false
        }
        return true
    }

    private static func findToken(
        _ chars: [Character],
        token: String,
        from: Int,
        end: Int
    ) -> Int? {
        let tokenChars = Array(token)
        guard tokenChars.isEmpty == false else { return nil }
        var index = from
        while index + tokenChars.count <= end {
            if chars[index] == "\\" {
                index += 2
                continue
            }
            if match(chars, index: index, end: end, token: token) {
                return index
            }
            index += 1
        }
        return nil
    }

    private static func canOpenUnderscore(_ chars: [Character], index: Int, end: Int) -> Bool {
        guard index < end, chars[index] == "_" else { return false }
        if index > 0, chars[index - 1].isLetter || chars[index - 1].isNumber {
            return false
        }
        return true
    }

    private static func serializeInlines(_ runs: [NoteInlineRun]) -> String {
        merge(runs).map { run in
            var text = escape(run.text)
            if run.bold {
                text = "**\(text)**"
            }
            if run.italic {
                text = "*\(text)*"
            }
            if run.underline {
                text = "<u>\(text)</u>"
            }
            return text
        }.joined()
    }

    private static func escape(_ text: String) -> String {
        var output = ""
        for character in text {
            switch character {
            case "\\", "*", "_", "<":
                output.append("\\")
                output.append(character)
            default:
                output.append(character)
            }
        }
        return output
    }

    private static func merge(_ runs: [NoteInlineRun]) -> [NoteInlineRun] {
        var merged: [NoteInlineRun] = []
        for run in runs {
            if run.text.isEmpty { continue }
            if let last = merged.last,
               last.bold == run.bold,
               last.italic == run.italic,
               last.underline == run.underline
            {
                merged[merged.count - 1].text += run.text
            } else {
                merged.append(run)
            }
        }
        return merged
    }
}

File: apps/web/lib/notes/note-markdown.ts
/**
 * Shared notes markdown subset with the iOS NoteMarkdown parser and TipTap toolbar:
 * bold (`**` / `__`), italic (`*` / `_`), underline (`<u>`), H1, H2, bullets.
 * Speaker labels stay as `## Me` / `## Speaker 1` headings.
 */

export type NoteMarkdownMarks = {
  bold: boolean;
  italic: boolean;
  underline: boolean;
};

export type NoteMarkdownRun = NoteMarkdownMarks & {
  text: string;
};

export type NoteMarkdownBlockKind =
  | 'paragraph'
  | 'heading1'
  | 'heading2'
  | 'bullet';

export type NoteMarkdownBlock = {
  kind: NoteMarkdownBlockKind;
  runs: NoteMarkdownRun[];
};

export function parseNoteMarkdown(markdown: string): NoteMarkdownBlock[] {
  const normalized = markdown.replace(/\r\n/g, '\n').replace(/\r/g, '\n');
  if (!normalized.trim()) {
    return [];
  }

  const blocks: NoteMarkdownBlock[] = [];
  for (const line of normalized.split('\n')) {
    if (!line.trim()) {
      continue;
    }

    if (line.startsWith('## ')) {
      blocks.push({
        kind: 'heading2',
        runs: parseInlines(line.slice(3)),
      });
    } else if (line.startsWith('# ')) {
      blocks.push({
        kind: 'heading1',
        runs: parseInlines(line.slice(2)),
      });
    } else {
      const item = bulletItem(line);
      if (item !== null) {
        blocks.push({ kind: 'bullet', runs: parseInlines(item) });
      } else {
        blocks.push({ kind: 'paragraph', runs: parseInlines(line) });
      }
    }
  }

  return blocks;
}

export function noteMarkdownToHtml(markdown: string): string {
  const blocks = parseNoteMarkdown(markdown);
  if (blocks.length === 0) {
    return '';
  }

  const parts: string[] = [];
  let index = 0;

  while (index < blocks.length) {
    const block = blocks[index]!;
    if (block.kind === 'bullet') {
      const items: string[] = [];
      while (index < blocks.length && blocks[index]!.kind === 'bullet') {
        items.push(`<li>${renderRuns(blocks[index]!.runs)}</li>`);
        index += 1;
      }
      parts.push(`<ul>${items.join('')}</ul>`);
      continue;
    }

    const inner = renderRuns(block.runs);
    if (block.kind === 'heading1') {
      parts.push(`<h1>${inner}</h1>`);
    } else if (block.kind === 'heading2') {
      parts.push(`<h2>${inner}</h2>`);
    } else {
      parts.push(`<p>${inner}</p>`);
    }
    index += 1;
  }

  return parts.join('');
}

export function noteMarkdownToPlainText(markdown: string): string {
  return parseNoteMarkdown(markdown)
    .map((block) => block.runs.map((run) => run.text).join(''))
    .map((text) => text.trim())
    .filter(Boolean)
    .join('\n');
}

function bulletItem(line: string): string | null {
  const rest = line.replace(/^ +/, '');
  for (const prefix of ['- ', '* ', '+ ', '• ']) {
    if (rest.startsWith(prefix)) {
      return rest.slice(prefix.length);
    }
  }
  return null;
}

function parseInlines(input: string): NoteMarkdownRun[] {
  return mergeRuns(
    parseInlinesRange(input, 0, input.length, {
      bold: false,
      italic: false,
      underline: false,
    }).runs,
  );
}

function parseInlinesRange(
  input: string,
  start: number,
  end: number,
  marks: NoteMarkdownMarks,
): { runs: NoteMarkdownRun[]; index: number } {
  const runs: NoteMarkdownRun[] = [];
  let buffer = '';
  let index = start;

  const flush = () => {
    if (!buffer) return;
    runs.push({ text: buffer, ...marks });
    buffer = '';
  };

  while (index < end) {
    if (input[index] === '\\' && index + 1 < end) {
      buffer += input[index + 1];
      index += 2;
      continue;
    }

    if (
      input.startsWith('<u>', index) &&
      index + 3 < end
    ) {
      const close = findToken(input, '</u>', index + 3, end);
      if (close !== -1) {
        flush();
        const nested = parseInlinesRange(input, index + 3, close, {
          ...marks,
          underline: true,
        });
        runs.push(...nested.runs);
        index = close + 4;
        continue;
      }
    }

    const wrapped =
      wrap(input, index, end, '***', marks, { bold: true, italic: true }) ??
      wrap(input, index, end, '___', marks, { bold: true, italic: true }) ??
      wrap(input, index, end, '**', marks, { bold: true }) ??
      wrap(input, index, end, '__', marks, { bold: true }) ??
      wrap(input, index, end, '*', marks, { italic: true }) ??
      (canOpenUnderscore(input, index, end)
        ? wrap(input, index, end, '_', marks, { italic: true })
        : null);

    if (wrapped) {
      flush();
      runs.push(...wrapped.runs);
      index = wrapped.index;
      continue;
    }

    buffer += input[index];
    index += 1;
  }

  flush();
  return { runs, index };
}

function wrap(
  input: string,
  index: number,
  end: number,
  delimiter: string,
  marks: NoteMarkdownMarks,
  apply: Partial<NoteMarkdownMarks>,
): { runs: NoteMarkdownRun[]; index: number } | null {
  if (!input.startsWith(delimiter, index)) {
    return null;
  }

  const innerStart = index + delimiter.length;
  const close = findToken(input, delimiter, innerStart, end);
  if (close === -1 || close <= innerStart) {
    return null;
  }

  const nested = parseInlinesRange(input, innerStart, close, {
    ...marks,
    ...apply,
  });
  return { runs: nested.runs, index: close + delimiter.length };
}

function findToken(
  input: string,
  token: string,
  from: number,
  end: number,
): number {
  let index = from;
  while (index + token.length <= end) {
    if (input[index] === '\\') {
      index += 2;
      continue;
    }
    if (input.startsWith(token, index)) {
      return index;
    }
    index += 1;
  }
  return -1;
}

function canOpenUnderscore(
  input: string,
  index: number,
  end: number,
): boolean {
  if (index >= end || input[index] !== '_') {
    return false;
  }
  if (index > 0 && /[A-Za-z0-9]/.test(input[index - 1]!)) {
    return false;
  }
  return true;
}

function mergeRuns(runs: NoteMarkdownRun[]): NoteMarkdownRun[] {
  const merged: NoteMarkdownRun[] = [];
  for (const run of runs) {
    if (!run.text) continue;
    const last = merged[merged.length - 1];
    if (
      last &&
      last.bold === run.bold &&
      last.italic === run.italic &&
      last.underline === run.underline
    ) {
      last.text += run.text;
    } else {
      merged.push({ ...run });
    }
  }
  return merged;
}

function renderRuns(runs: NoteMarkdownRun[]): string {
  return runs
    .map((run) => {
      let html = escapeHtml(run.text);
      if (run.bold) html = `<strong>${html}</strong>`;
      if (run.italic) html = `<em>${html}</em>`;
      if (run.underline) html = `<u>${html}</u>`;
      return html;
    })
    .join('');
}

function escapeHtml(text: string): string {
  return text
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;');
}
